diff --git a/mojoshader.c b/mojoshader.c
index 68ee5be..d57769d 100644
--- a/mojoshader.c
+++ b/mojoshader.c
@@ -3047,7 +3047,8 @@ static void state_SINCOS(Context *ctx)
 
 static void state_IF(Context *ctx)
 {
-    const RegisterType regtype = ctx->source_args[0].regtype;
+    const SourceArgInfo *src = &ctx->source_args[0];
+    const RegisterType regtype = src->regtype;
     if ((regtype != REG_TYPE_PREDICATE) && (regtype != REG_TYPE_CONSTBOOL))
     {
         fail(ctx, "IF src0 must be CONSTBOOL or PREDICATE");
@@ -3059,6 +3060,18 @@ static void state_IF(Context *ctx)
         fail(ctx, "IF predicate argument requires Shader Model 2.x or later");
         return;
     } // else if
+    else if ((src->src_mod != SRCMOD_NONE) &&
+             (src->src_mod != SRCMOD_NOT))
+    {
+        fail(ctx, "IF source modifier must be NONE or NOT");
+        return;
+    } // else if
+    else if ((regtype == REG_TYPE_PREDICATE) &&
+             !replicate_swizzle(src->swizzle))
+    {
+        fail(ctx, "IF predicate source must have replicate swizzle");
+        return;
+    } // else if
 
     const int dynamic = (regtype == REG_TYPE_PREDICATE);
     if (!push_flow_control_block(ctx, dynamic ? FLOW_CONTROL_BLOCK_IF_DYNAMIC
