diff --git a/mojoshader.c b/mojoshader.c
--- a/mojoshader.c
+++ b/mojoshader.c
@@ -1975,7 +1975,14 @@ static void state_TEXLD(Context *ctx)
 
     else if (shader_version_atleast(ctx, 1, 4))
     {
-        // !!! FIXME: checks for ps_1_4 version here...
+        const DestArgInfo *dst = &ctx->dest_arg;
+        const SourceArgInfo *src = &ctx->source_args[0];
+        if (dst->regtype != REG_TYPE_TEMP)
+            fail(ctx, "ps_1_4 TEXLD destination must be a temp register");
+        if ((src->regtype != REG_TYPE_TEXTURE) &&
+            (src->regtype != REG_TYPE_TEMP))
+            fail(ctx, "ps_1_4 TEXLD source must be a texture or temp register");
+        add_sampler(ctx, dst->regnum, TEXTURE_TYPE_2D, 0);
     } // else if
 
     else
diff --git a/profiles/mojoshader_profile_glsl.c b/profiles/mojoshader_profile_glsl.c
--- a/profiles/mojoshader_profile_glsl.c
+++ b/profiles/mojoshader_profile_glsl.c
@@ -2085,9 +2085,46 @@ void emit_GLSL_TEXLD(Context *ctx)
 
     else if (!shader_version_atleast(ctx, 2, 0))
     {
-        // ps_1_4 is different, too!
-        fail(ctx, "TEXLD == Shader Model 1.4 unimplemented.");  // !!! FIXME
-        return;
+        const DestArgInfo *info = &ctx->dest_arg;
+        RegisterList *sreg = reglist_find(&ctx->samplers, REG_TYPE_SAMPLER,
+                                          info->regnum);
+        if (sreg == NULL)
+        {
+            fail(ctx, "ps_1_4 TEXLD using undeclared sampler");
+            return;
+        } // if
+
+        char sampler[64];
+        get_GLSL_varname_in_buf(ctx, REG_TYPE_SAMPLER, info->regnum,
+                                sampler, sizeof (sampler));
+        char coordinates[256];
+        const char *funcname = NULL;
+        switch ((const TextureType) sreg->index)
+        {
+            case TEXTURE_TYPE_2D:
+                funcname = "texture2D";
+                make_GLSL_srcarg_string_vec2(ctx, 0, coordinates,
+                                             sizeof (coordinates));
+                break;
+            case TEXTURE_TYPE_CUBE:
+                funcname = "textureCube";
+                make_GLSL_srcarg_string_vec3(ctx, 0, coordinates,
+                                             sizeof (coordinates));
+                break;
+            case TEXTURE_TYPE_VOLUME:
+                funcname = "texture3D";
+                make_GLSL_srcarg_string_vec3(ctx, 0, coordinates,
+                                             sizeof (coordinates));
+                break;
+            default:
+                fail(ctx, "unknown ps_1_4 texture type");
+                return;
+        } // switch
+
+        char code[256];
+        make_GLSL_destarg_assign(ctx, code, sizeof (code), "%s(%s, %s)",
+                                 funcname, sampler, coordinates);
+        output_line(ctx, "%s", code);
     } // else if
 
     else
