diff --git a/mojoshader.c b/mojoshader.c
--- a/mojoshader.c
+++ b/mojoshader.c
@@ -1473,7 +1473,11 @@ static int parse_source_token(Context *ctx, SourceArgInfo *info)
             ctx->have_dest_arg && (ctx->dest_arg.regtype == REG_TYPE_TEMP) &&
             (ctx->dest_arg.regnum == info->regnum) &&
             !ctx->dest_arg_was_written;
-        if (!ctx->ignore_temp_read_validation &&
+        const int validates_uninitialized_temporaries =
+            (shader_is_pixel(ctx) && !shader_version_atleast(ctx, 3, 0)) ||
+            (shader_is_vertex(ctx) && !shader_version_atleast(ctx, 2, 0));
+        if (validates_uninitialized_temporaries &&
+            !ctx->ignore_temp_read_validation &&
             (info->regtype == REG_TYPE_TEMP) && (reg) &&
             ((!reg->written) || uninitialized_destination_self_read))
             failf(ctx, "Temp register r%d used uninitialized", info->regnum);
