diff --git a/mojoshader.c b/mojoshader.c
--- a/mojoshader.c
+++ b/mojoshader.c
@@ -1106,7 +1106,7 @@ static int parse_args_DCL(Context *ctx)
             const uint32 ttype = ((token >> 27) & 0xF);
             if (!valid_texture_type(ttype))
                 fail(ctx, "Unknown sampler texture type");
-            reserved_mask = 0x0FFFFFFF;
+            reserved_mask = 0x07FFFFFF;
             ctx->dwords[0] = ttype;
         } // else if
         else
