diff --git a/mojoshader.c b/mojoshader.c
--- a/mojoshader.c
+++ b/mojoshader.c
@@ -3265,4 +3265,7 @@ static void state_TEXLD(Context *ctx)
         else if (src->src_mod != SRCMOD_NONE)
             fail(ctx, "ps_1_4 TEXLD source modifier must be NONE, DZ, or DW");
+        else if ((src->regtype == REG_TYPE_TEMP) &&
+                 (src->swizzle != 0xE4) && (src->swizzle != 0xA4))
+            fail(ctx, "ps_1_4 TEXLD temporary source must select XYZ");
         if (src->regtype == REG_TYPE_TEXTURE)
             validate_pixel14_texture_coordinate_selector(ctx, src);
