diff --git a/mojoshader.c b/mojoshader.c
--- a/mojoshader.c
+++ b/mojoshader.c
@@ -1630,7 +1630,9 @@ static int parse_source_token(Context *ctx, SourceArgInfo *info)
             (ctx->dest_arg.regnum == info->regnum) &&
             !ctx->dest_arg_was_written;
         const int validates_uninitialized_temporaries =
-            (shader_is_pixel(ctx) && !shader_version_atleast(ctx, 3, 0)) ||
+            (shader_is_pixel(ctx) &&
+             (!shader_version_atleast(ctx, 2, 0) ||
+              shader_version_exactly(ctx, 2, 0))) ||
             (shader_is_vertex(ctx) && !shader_version_atleast(ctx, 2, 0));
         if (validates_uninitialized_temporaries &&
             !ctx->ignore_temp_read_validation &&
@@ -2384,8 +2386,10 @@ static void srcarg_matrix_replicate(Context *ctx, const int idx,
         set_used_register(ctx, dst->regtype, dst->regnum, 0);
     } // for
 
     const int validates_uninitialized_temporaries =
-        (shader_is_pixel(ctx) && !shader_version_atleast(ctx, 3, 0)) ||
+        (shader_is_pixel(ctx) &&
+         (!shader_version_atleast(ctx, 2, 0) ||
+          shader_version_exactly(ctx, 2, 0))) ||
         (shader_is_vertex(ctx) && !shader_version_atleast(ctx, 2, 0));
     if (!validates_uninitialized_temporaries ||
         ctx->ignore_temp_read_validation)
@@ -3075,6 +3079,9 @@ static void state_TEXKILL(Context *ctx)
     const DestArgInfo *info = &ctx->dest_arg;
     const RegisterType regtype = info->regtype;
     const int shader_model_three = shader_version_atleast(ctx, 3, 0);
+    const int validates_uninitialized_temporary =
+        !shader_version_atleast(ctx, 2, 0) ||
+        shader_version_exactly(ctx, 2, 0);
     if (!writemask_xyzw(info->writemask))
         fail(ctx, "TEXKILL writemask must be .xyzw");
     else if (!shader_version_atleast(ctx, 1, 4) &&
@@ -3086,7 +3093,7 @@ static void state_TEXKILL(Context *ctx)
     else if (!shader_model_three && shader_version_atleast(ctx, 1, 4) &&
              (regtype != REG_TYPE_TEMP) && (regtype != REG_TYPE_TEXTURE))
         fail(ctx, "TEXKILL must use a temp or texture register");
-    else if (!shader_model_three && (regtype == REG_TYPE_TEMP) &&
+    else if (validates_uninitialized_temporary && (regtype == REG_TYPE_TEMP) &&
              ((ctx->dest_arg_prior_writemask & 0x7) != 0x7))
         fail(ctx, "TEXKILL temporary XYZ components are uninitialized");
     else if (shader_version_exactly(ctx, 1, 4) && (ctx->phase_count == 0))
