diff --git a/mojoshader.c b/mojoshader.c
--- a/mojoshader.c
+++ b/mojoshader.c
@@ -3069,6 +3069,8 @@ static void state_TEXLDL(Context *ctx)
 {
     if (!shader_version_atleast(ctx, 3, 0))
         fail(ctx, "TEXLDL in version < Shader Model 3.0");
+    else if (ctx->dest_arg.result_mod & MOD_SATURATE)
+        fail(ctx, "TEXLDL destination must not saturate");
     else if (ctx->source_args[0].src_mod != SRCMOD_NONE)
         fail(ctx, "TEXLDL src0 must have no modifiers");
     else if (ctx->source_args[1].regtype != REG_TYPE_SAMPLER)
